tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,647 @@
+//
+// Note that it is assumed that the provided database resource has at least one
+// matching names as they are retrieved from the backend based on one of those
+// names.
+func ensureEachDatabase(names []string, dbs []types.Database) error {
+	byDiscoveredNameOrName := map[string]types.Databases{}
+	for _, db := range dbs {
+		// Database may be listed by their original name in the cloud.
+		byDiscoveredNameOrName[db.GetName()] = append(byDiscoveredNameOrName[db.GetName()], db)
+
+		if discoveredName, ok := common.GetDiscoveredResourceName(db); ok && discoveredName != db.GetName() {
+			byDiscoveredNameOrName[discoveredName] = append(byDiscoveredNameOrName[discoveredName], db)
+		}
+	}
+
+	for _, name := range names {
+		matched := byDiscoveredNameOrName[name]
+		switch len(matched) {
+		case 0:
+			return trace.NotFound("database %q not found", name)
+		case 1:
+			continue
+		default:
+			var sb strings.Builder
+			printTableForDatabaseExec(&sb, matched)
+			return trace.BadParameter(`%q matches multiple databases:
+%vTry selecting the database with a more specific name printed in the above table`, name, sb.String())
+		}
+	}
+
+	return nil
+}
+
+func printTableForDatabaseExec(w io.Writer, dbs []types.Database) {
+	rows := make([]databaseTableRow, 0, len(dbs))
+	for _, db := range dbs {
+		// Always use full name but don't print hidden labels.
+		row := getDatabaseRow("", "", "", db, nil, nil, false)
+		row.DisplayName = db.GetName()
+		rows = append(rows, row)
+	}
+	printDatabaseTable(printDatabaseTableConfig{
+		writer:         w,
+		rows:           rows,
+		includeColumns: []string{"Name", "Protocol", "Description", "Labels"},
+	})
+}
+
+type databaseExecResult struct {
+	proto.RouteToDatabase `json:"database"`
+	Command               string `json:"command"`
+	OutputFile            string `json:"output_file,omitempty"`
+	Success               bool   `json:"success"`
+	Error                 string `json:"error,omitempty"`
+	ExitCode              int    `json:"exit_code"`
+}
+
+type databaseExecSummary struct {
+	Databases []databaseExecResult `json:"databases"`
+	Success   int                  `json:"success"`
+	Failure   int                  `json:"failure"`
+	Total     int                  `json:"total"`
+
+	mu sync.Mutex
+}
+
+func (s *databaseExecSummary) add(result databaseExecResult) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.Databases = append(s.Databases, result)
+	s.Total++
+	if result.Success {
+		s.Success++
+	} else {
+		s.Failure++
+	}
+}
+
+func (s *databaseExecSummary) print(w io.Writer) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	fmt.Fprintf(w, "\nSummary: %d of %d succeeded.\n", s.Success, s.Total)
+}
+
+func (s *databaseExecSummary) printAndSave(w io.Writer, outputDir string) {
+	s.print(w)
+	if err := s.save(w, outputDir); err != nil {
+		fmt.Fprintf(w, "Failed to save summary: %v\n", err)
+	}
+}
+
+func (s *databaseExecSummary) save(w io.Writer, outputDir string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
It is generally not recommended to hold a lock over I/O. From what I can see it's only needed to protect `s`, can we scope this down to only when we actually need it rather then entire function?